cli/src/cmd/cast/logs.rs

```diff
@@ -0,0 +1,405 @@
+// cast estimate subcommands
+use crate::{
+    opts::EthereumOpts,
+    utils::{self},
+};
+use cast::Cast;
+use clap::Parser;
+use ethers::{
+    abi::{Address, RawTopicFilter, Topic, TopicFilter},
+    providers::Middleware,
+    types::{BlockId, BlockNumber, Filter, FilterBlockOption, NameOrAddress, ValueOrArray, H256},
+};
+
+use foundry_common::abi::{get_event, parse_tokens};
+use foundry_config::Config;
+
+use itertools::Itertools;
+
+use std::str::FromStr;
+
+/// CLI arguments for `cast access-list`.
+#[derive(Debug, Parser)]
+pub struct LogsArgs {
+    #[clap(flatten)]
+    eth: EthereumOpts,
+    /// The block height to start query at.
+    ///
+    /// Can also be the tags earliest, finalized, safe, latest, or pending.
+    #[clap(long)]
+    from_block: Option<BlockId>,
+
+    /// The block height to stop query at.
+    ///
+    /// Can also be the tags earliest, finalized, safe, latest, or pending.
+    #[clap(long)]
+    to_block: Option<BlockId>,
+
+    /// The contract address to filter on.
+    #[clap(
+        long,
+        value_parser = NameOrAddress::from_str
+    )]
+    address: Option<NameOrAddress>,
+
+    /// The signature of the event to filter logs by which will be converted to the first topic or
+    /// a topic to filter on.
+    #[clap(value_name = "SIG_OR_TOPIC")]
+    sig_or_topic: Option<String>,
+
+    /// If used with a signature, the indexed fields of the event to filter by. Otherwise, the
+    /// remaining topics of the filter.
+    #[clap(value_name = "TOPICS_OR_ARGS")]
+    topics_or_args: Vec<String>,
+
+    /// Print the logs as JSON.
+    #[clap(long, short, help_heading = "Display options")]
+    json: bool,
+}
+
+impl LogsArgs {
+    pub async fn run(self) -> eyre::Result<()> {
+        let LogsArgs {
+            from_block, to_block, address, topics_or_args, sig_or_topic, json, eth, ..
+        } = self;
+
+        let config = Config::from(&eth);
+        let provider = utils::get_provider(&config)?;
+
+        let address = match address {
+            Some(address) => {
+                let address = match address {
+                    NameOrAddress::Name(name) => provider.resolve_name(&name).await?,
+                    NameOrAddress::Address(address) => address,
+                };
+                Some(address)
+            }
+            None => None,
+        };
+
+        let from_block = get_block_number(&provider, from_block).await?;
+        let to_block = get_block_number(&provider, to_block).await?;
+
+        let cast = Cast::new(&provider);
+
+        let filter = build_filter(from_block, to_block, address, sig_or_topic, topics_or_args)?;
+
+        let logs = cast.filter_logs(filter, json).await?;
+
+        println!("{}", logs);
+
+        Ok(())
+    }
+}
+
+async fn get_block_number<M: Middleware>(
```

Comment:
would like a one line here what this does: converting the BlockId

Reply:
Renamed to `convert_block_number` to prevent overlap with existing functions in the repo. Also added a bit of documentation.